documentation/proposals/Proposal - Axis Input Devices.md

```diff
@@ -0,0 +1,624 @@
+        new(37, AxisTrait.EulerAngleComponent, "Gyro Y"),
+        new(38, AxisTrait.EulerAngleComponent, "Gyro Z"),
+
+        // accelerometer
+        new(39, AxisTrait.Acceleration, "Accelerometer X"),
+        new(40, AxisTrait.Acceleration, "Accelerometer Y"),
+        new(41, AxisTrait.Acceleration, "Accelerometer Z"),
+
+        // battery level
+        new(42, AxisTrait.Analog, "Battery Level"),
+
+        // microphone input level
+        new(43, AxisTrait.Analog, "Microphone Level"),
+    ];
+}
+ ```
+
+As additional food for thought, a Sony Dualshock 2 controller whose face buttons are famously (though infrequently used as) analog, might contain face buttons with the following definition:
+
+```csharp
+_ = new AxisDescription(0, AxisTrait.Binary | AxisTrait.Analog | AxisTrait.RightSide, "Square");
+```
+
+## Axis Groups
+`AxisGroup`s refer to collections of axes whose outputs or layouts are related to one another. If used smartly, they can be used to provide novel high-level remapping schemes, virtual devices, or other features.
+
+### Struct Definition
+
+```csharp
+/// <summary>
+/// This is the declaration of a group of axes within a specific device
+/// </summary>
+/// <param name="Index">Index within the list of <see cref="AxisGroup"/>s</param>
+/// <param name="Purpose">The intended purpose of this group</param>
+/// <param name="Axes">The indices of the axes that make up this group</param>
+/// <param name="TwinIndex">The AxisGroup index of the "twin" of this group - e.g. if this is the left thumbstick, the twin would be the right thumbstick</param>
+/// <param name="Name">A human-readable name for the group</param>
+/// <remarks>Can names be inferred at runtime from the AxisDescription and AxisGroup? Should they be nullable/optional
+/// in the definition and then populated at runtime if they're missing?</remarks>
+public readonly record struct AxisGroup(
+    int Index,
+    AxisGroupType Purpose,
+    IReadOnlyList<int> Axes,
+    int? TwinIndex = null,
+    string Name = "Unknown")
+{
+    public static implicit operator int(AxisGroup group) => group.Index;
+}
+```
+
+### Axis Group Types
+`AxisGroupType` is the primary way in which an `AxisGroup` is given meaning. It is how analog sticks, touchpads, DPads, mouse travel, etc are defined. 
+
+Like `AxisDescription`, it contains hard-coded constraints to help ensure that groups are defined correctly. It also has a few flags that enforce certain requirements on the implementer. We should strive to provide default implementations of expected behaviours regarding these constraints wherever possible. This is another important source of information that must be validated at runtime.
+
+
+
+```csharp
+/// <summary>
+/// This is how one defines the purpose of groups of multiple axes in a device
+/// It has specifications that must be followed to be considered valid
+/// </summary>
+[Flags]
+public enum AxisGroupType : uint

```

Comment:
```suggestion
public enum AxisGroupType : ulong

```

Reply:
ultimately i have no issue with this whatsoever,  but I think there is plenty of room as-is, and it might be good to discourage over-populating this enum. that being said, we all know this is going to need additions at *some* point in the future, so I have no strong feelings.

Reply:
fuck it i'm shipping the ulong bc i got some evil ideas for the future